File: internal/eventstore/storage.go
package eventstore

import (
	"context"
	"fmt"
	"sync"
)

// Storage defines the interface for event storage backends
// This abstraction allows different storage engines (SQLite, PostgreSQL, RocksDB, LevelDB)
type Storage interface {
	// Lifecycle methods
	Open(ctx context.Context, config map[string]interface{}) error
	Close(ctx context.Context) error

	// Event operations
	WriteEvent(ctx context.Context, event Event) error
	ReadEvents(ctx context.Context, aggregateID string, fromSeq int64, limit int) ([]Event, error)
	GetLastSequence(ctx context.Context, aggregateID string) (int64, error)

	// Snapshot operations
	WriteSnapshot(ctx context.Context, aggregateID string, sequence int64, data []byte) error
	ReadSnapshot(ctx context.Context, aggregateID string) (sequence int64, data []byte, err error)

	// Maintenance operations
	Compact(ctx context.Context) error
	GetStats(ctx context.Context) (map[string]interface{}, error)

	// Batch operations for performance
	WriteEventsBatch(ctx context.Context, events []Event) error
}

// StorageConfig contains configuration for storage backends
type StorageConfig struct {
	Type             string                 `yaml:"type"` // sqlite, postgresql, rocksdb, leveldb
	Path             string                 `yaml:"path,omitempty"`
	ConnectionString string                 `yaml:"connection_string,omitempty"`
	Options          map[string]interface{} `yaml:"options,omitempty"`
}

// StorageFactory creates storage instances based on type
type StorageFactory struct{}

// NewStorage creates a storage instance based on configuration
func (f *StorageFactory) NewStorage(config StorageConfig) (Storage, error) {
	switch config.Type {
	case "sqlite":
		return NewSQLiteStorage(), nil
	case "postgresql":
		return NewPostgreSQLStorage(), nil
	case "memory":
		return NewMemoryStorage(), nil
	default:
		return NewMemoryStorage(), nil // Default to memory for MVP
	}
}

// MemoryStorage implements in-memory storage for MVP/testing
type MemoryStorage struct {
	events    map[string][]Event
	snapshots map[string]SnapshotData
	sequences map[string]int64
	mu        sync.RWMutex
}

type SnapshotData struct {
	Sequence int64
	Data     []byte
}

// NewMemoryStorage creates a new in-memory storage
func NewMemoryStorage() *MemoryStorage {
	return &MemoryStorage{
		events:    make(map[string][]Event),
		snapshots: make(map[string]SnapshotData),
		sequences: make(map[string]int64),
	}
}

// Open implements Storage
func (m *MemoryStorage) Open(ctx context.Context, config map[string]interface{}) error {
	return nil
}

// Close implements Storage
func (m *MemoryStorage) Close(ctx context.Context) error {
	return nil
}

// WriteEvent implements Storage
func (m *MemoryStorage) WriteEvent(ctx context.Context, event Event) error {
	m.mu.Lock()
	defer m.mu.Unlock()

	if events, exists := m.events[event.AggregateID]; exists {
		// Check sequence
		if len(events) > 0 && event.Sequence != events[len(events)-1].Sequence+1 {
			return fmt.Errorf("sequence mismatch for aggregate %s", event.AggregateID)
		}
	} else {
		// First event for this aggregate
		if event.Sequence != 1 {
			return fmt.Errorf("first event must have sequence 1 for aggregate %s", event.AggregateID)
		}
	}

	m.events[event.AggregateID] = append(m.events[event.AggregateID], event)
	m.sequences[event.AggregateID] = event.Sequence

	return nil
}

// ReadEvents implements Storage
func (m *MemoryStorage) ReadEvents(ctx context.Context, aggregateID string, fromSeq int64, limit int) ([]Event, error) {
	m.mu.RLock()
	defer m.mu.RUnlock()

	events, exists := m.events[aggregateID]
	if !exists {
		return []Event{}, nil
	}

	var result []Event
	for _, event := range events {
		if event.Sequence >= fromSeq {
			result = append(result, event)
			if limit > 0 && len(result) >= limit {
				break
			}
		}
	}

	return result, nil
}

// GetLastSequence implements Storage
func (m *MemoryStorage) GetLastSequence(ctx context.Context, aggregateID string) (int64, error) {
	m.mu.RLock()
	defer m.mu.RUnlock()

	if seq, exists := m.sequences[aggregateID]; exists {
		return seq, nil
	}
	return 0, nil
}

// WriteSnapshot implements Storage
func (m *MemoryStorage) WriteSnapshot(ctx context.Context, aggregateID string, sequence int64, data []byte) error {
	m.mu.Lock()
	defer m.mu.Unlock()

	m.snapshots[aggregateID] = SnapshotData{
		Sequence: sequence,
		Data:     data,
	}
	return nil
}

// ReadSnapshot implements Storage
func (m *MemoryStorage) ReadSnapshot(ctx context.Context, aggregateID string) (int64, []byte, error) {
	m.mu.RLock()
	defer m.mu.RUnlock()

	if snapshot, exists := m.snapshots[aggregateID]; exists {
		return snapshot.Sequence, snapshot.Data, nil
	}
	return 0, nil, nil
}

// Compact implements Storage
func (m *MemoryStorage) Compact(ctx context.Context) error {
	return nil
}

// GetStats implements Storage
func (m *MemoryStorage) GetStats(ctx context.Context) (map[string]interface{}, error) {
	m.mu.RLock()
	defer m.mu.RUnlock()

	return map[string]interface{}{
		"total_aggregates": len(m.events),
		"total_events":     m.countTotalEvents(),
		"memory_usage":     "unknown",
	}, nil
}

func (m *MemoryStorage) countTotalEvents() int {
	count := 0
	for _, events := range m.events {
		count += len(events)
	}
	return count
}

// WriteEventsBatch implements Storage
func (m *MemoryStorage) WriteEventsBatch(ctx context.Context, events []Event) error {
	for _, event := range events {
		if err := m.WriteEvent(ctx, event); err != nil {
			return err
		}
	}
	return nil
}

// GetAllEvents returns all events (for testing/debugging)
func (m *MemoryStorage) GetAllEvents() []Event {
	m.mu.RLock()
	defer m.mu.RUnlock()

	var allEvents []Event
	for _, events := range m.events {
		allEvents = append(allEvents, events...)
	}
	return allEvents
}

File: demo/nextjs-supabase/src/lib/actordb/eventstore/storage.ts
import { EventStoreConfig } from '../config/config';

// Event represents a single event in the event store
// Merkle DAG: sha256:event_v1 - Single-writer actor event persistence
export interface Event {
  aggregateId: string;
  sequence: number;
  eventType: string;
  data: Buffer;
  timestamp: Date;
  eventTime: Date; // For late event handling
  aggregateType: string;
  metadata?: { [key: string]: any };
}

// ActorState represents the current state of an actor
export interface ActorState {
  aggregateId: string;
  lastSequence: number;
  lastTimestamp: Date;
  snapshotData?: Buffer;
  snapshotSeq?: number;
}

// WriteResult contains the result of a write operation
export interface WriteResult {
  aggregateId: string;
  sequence: number;
  timestamp: Date;
  success: boolean;
  error?: string;
}

// Storage defines the interface for event storage backends
// Process Network Node: storage_backend
// Dependencies: []
// Outputs: [persisted_events, snapshots]
// Merkle DAG: sha256:storage_interface_v1 - Abstract storage backend interface
// This abstraction allows different storage engines (SQLite, PostgreSQL, RocksDB, LevelDB)
export interface Storage {
  // Lifecycle methods
  open(config: { [key: string]: any }): Promise<void>;
  close(): Promise<void>;

  // Event operations
  writeEvent(event: Event): Promise<void>;
  readEvents(aggregateId: string, fromSeq: number, limit?: number): Promise<Event[]>;
  getLastSequence(aggregateId: string): Promise<number>;

  // Snapshot operations
  writeSnapshot(aggregateId: string, sequence: number, data: Buffer): Promise<void>;
  readSnapshot(aggregateId: string): Promise<{ sequence: number; data: Buffer } | null>;

  // Maintenance operations
  compact(): Promise<void>;
  getStats(): Promise<{ [key: string]: any }>;

  // Batch operations for performance
  writeEventsBatch(events: Event[]): Promise<void>;
}

// StorageConfig contains configuration for storage backends
export interface StorageConfig {
  type: string; // sqlite, postgresql, rocksdb, leveldb, memory
  path?: string;
  connectionString?: string;
  options?: { [key: string]: any };
}

// StorageFactory creates storage instances based on type
export class StorageFactory {
  static async createStorage(config: StorageConfig): Promise<Storage> {
    switch (config.type) {
      case 'sqlite':
        return new SQLiteStorage();
      case 'postgresql':
        return new PostgreSQLStorage();
      case 'memory':
      default:
        return new MemoryStorage();
    }
  }
}

// MemoryStorage implements in-memory storage for MVP/testing
// Merkle DAG: sha256:memory_storage_v1 - In-memory event storage for development/testing
export class MemoryStorage implements Storage {
  private events: Map<string, Event[]> = new Map();
  private snapshots: Map<string, { sequence: number; data: Buffer }> = new Map();
  private sequences: Map<string, number> = new Map();

  async open(config: { [key: string]: any }): Promise<void> {
    // No-op for memory storage
  }

  async close(): Promise<void> {
    // No-op for memory storage
  }

  async writeEvent(event: Event): Promise<void> {
    const existingEvents = this.events.get(event.aggregateId) || [];

    // Check sequence
    if (existingEvents.length > 0 && event.sequence !== existingEvents[existingEvents.length - 1].sequence + 1) {
      throw new Error(`sequence mismatch for aggregate ${event.aggregateId}`);
    }

    // First event for this aggregate
    if (existingEvents.length === 0 && event.sequence !== 1) {
      throw new Error(`first event must have sequence 1 for aggregate ${event.aggregateId}`);
    }

    this.events.set(event.aggregateId, [...existingEvents, event]);
    this.sequences.set(event.aggregateId, event.sequence);
  }

  async readEvents(aggregateId: string, fromSeq: number, limit?: number): Promise<Event[]> {
    const events = this.events.get(aggregateId) || [];
    const result: Event[] = [];

    for (const event of events) {
      if (event.sequence >= fromSeq) {
        result.push(event);
        if (limit && result.length >= limit) {
          break;
        }
      }
    }

    return result;
  }

  async getLastSequence(aggregateId: string): Promise<number> {
    return this.sequences.get(aggregateId) || 0;
  }

  async writeSnapshot(aggregateId: string, sequence: number, data: Buffer): Promise<void> {
    this.snapshots.set(aggregateId, { sequence, data });
  }

  async readSnapshot(aggregateId: string): Promise<{ sequence: number; data: Buffer } | null> {
    return this.snapshots.get(aggregateId) || null;
  }

  async compact(): Promise<void> {
    // No-op for memory storage
  }

  async getStats(): Promise<{ [key: string]: any }> {
    return {
      totalAggregates: this.events.size,
      totalEvents: this.countTotalEvents(),
      memoryUsage: 'unknown',
    };
  }

  async writeEventsBatch(events: Event[]): Promise<void> {
    for (const event of events) {
      await this.writeEvent(event);
    }
  }

  // Helper method for testing
  getAllEvents(): Event[] {
    const allEvents: Event[] = [];
    for (const events of this.events.values()) {
      allEvents.push(...events);
    }
    return allEvents;
  }

  private countTotalEvents(): number {
    let count = 0;
    for (const events of this.events.values()) {
      count += events.length;
    }
    return count;
  }
}

// SQLiteStorage implements SQLite storage
export class SQLiteStorage implements Storage {
  async open(config: { [key: string]: any }): Promise<void> {
    // TODO: Implement SQLite storage
    throw new Error('SQLite storage not implemented yet');
  }

  async close(): Promise<void> {
    // TODO: Implement close
  }

  async writeEvent(event: Event): Promise<void> {
    // TODO: Implement writeEvent
    throw new Error('SQLite writeEvent not implemented yet');
  }

  async readEvents(aggregateId: string, fromSeq: number, limit?: number): Promise<Event[]> {
    // TODO: Implement readEvents
    throw new Error('SQLite readEvents not implemented yet');
  }

  async getLastSequence(aggregateId: string): Promise<number> {
    // TODO: Implement getLastSequence
    throw new Error('SQLite getLastSequence not implemented yet');
  }

  async writeSnapshot(aggregateId: string, sequence: number, data: Buffer): Promise<void> {
    // TODO: Implement writeSnapshot
    throw new Error('SQLite writeSnapshot not implemented yet');
  }

  async readSnapshot(aggregateId: string): Promise<{ sequence: number; data: Buffer } | null> {
    // TODO: Implement readSnapshot
    throw new Error('SQLite readSnapshot not implemented yet');
  }

  async compact(): Promise<void> {
    // TODO: Implement compact
  }

  async getStats(): Promise<{ [key: string]: any }> {
    // TODO: Implement getStats
    return {};
  }

  async writeEventsBatch(events: Event[]): Promise<void> {
    // TODO: Implement writeEventsBatch
    throw new Error('SQLite writeEventsBatch not implemented yet');
  }
}

// PostgreSQLStorage implements PostgreSQL storage
export class PostgreSQLStorage implements Storage {
  async open(config: { [key: string]: any }): Promise<void> {
    // TODO: Implement PostgreSQL storage
    throw new Error('PostgreSQL storage not implemented yet');
  }

  async close(): Promise<void> {
    // TODO: Implement close
  }

  async writeEvent(event: Event): Promise<void> {
    // TODO: Implement writeEvent
    throw new Error('PostgreSQL writeEvent not implemented yet');
  }

  async readEvents(aggregateId: string, fromSeq: number, limit?: number): Promise<Event[]> {
    // TODO: Implement readEvents
    throw new Error('PostgreSQL readEvents not implemented yet');
  }

  async getLastSequence(aggregateId: string): Promise<number> {
    // TODO: Implement getLastSequence
    throw new Error('PostgreSQL getLastSequence not implemented yet');
  }

  async writeSnapshot(aggregateId: string, sequence: number, data: Buffer): Promise<void> {
    // TODO: Implement writeSnapshot
    throw new Error('PostgreSQL writeSnapshot not implemented yet');
  }

  async readSnapshot(aggregateId: string): Promise<{ sequence: number; data: Buffer } | null> {
    // TODO: Implement readSnapshot
    throw new Error('PostgreSQL readSnapshot not implemented yet');
  }

  async compact(): Promise<void> {
    // TODO: Implement compact
  }

  async getStats(): Promise<{ [key: string]: any }> {
    // TODO: Implement getStats
    return {};
  }

  async writeEventsBatch(events: Event[]): Promise<void> {
    // TODO: Implement writeEventsBatch
    throw new Error('PostgreSQL writeEventsBatch not implemented yet');
  }
}
